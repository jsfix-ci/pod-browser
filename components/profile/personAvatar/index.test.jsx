/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useEffect } from "react";
import { render, screen } from "@testing-library/react";
import { waitFor } from "@testing-library/dom";
import { CombinedDataProvider } from "@inrupt/solid-ui-react";
import * as solidClientFns from "@inrupt/solid-client";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { renderWithTheme } from "../../../__testUtils/withTheme";
import {
  aliceWebIdUrl,
  mockPersonDatasetAlice,
} from "../../../__testUtils/mockPersonResource";
import ConfirmationDialogContext from "../../../src/contexts/confirmationDialogContext";

import mockSession from "../../../__testUtils/mockSession";
import mockSessionContextProvider from "../../../__testUtils/mockSessionContextProvider";
import PersonAvatar, {
  setupErrorComponent,
  TESTCAFE_ID_UPLOAD_IMAGE,
  TESTCAFE_ID_REMOVE_IMAGE,
  confirmationDialogTitle,
  openDeleteConfirmationDialogWrapper,
  deletePhotoFunction,
} from "./index";
import mockConfirmationDialogContextProvider from "../../../__testUtils/mockConfirmationDialogContextProvider";
import {
  TESTCAFE_ID_CONFIRMATION_DIALOG,
  TESTCAFE_ID_CONFIRMATION_DIALOG_TITLE,
} from "../../confirmationDialog";

const profileIri = "https://example.com/profile/card#me";

describe("Person Avatar", () => {
  const profileDataset = mockPersonDatasetAlice();
  const profileThing = solidClientFns.getThing(profileDataset, aliceWebIdUrl);

  beforeEach(() => {
    jest
      .spyOn(solidClientFns, "getSolidDataset")
      .mockResolvedValue(profileDataset);
    jest.spyOn(solidClientFns, "getThing").mockReturnValue(profileThing);
  });

  test("renders a person avatar", async () => {
    const session = mockSession();
    const SessionProvider = mockSessionContextProvider(session);
    const { asFragment, getByText } = renderWithTheme(
      <CombinedDataProvider solidDataset={profileDataset} thing={profileThing}>
        <SessionProvider>
          <PersonAvatar profileIri={profileIri} />
        </SessionProvider>
      </CombinedDataProvider>
    );
    await waitFor(() => {
      expect(getByText("Alice")).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders an upload/change button ", async () => {
    const session = mockSession();
    const SessionProvider = mockSessionContextProvider(session);

    const { getByTestId } = renderWithTheme(
      <CombinedDataProvider solidDataset={profileDataset} thing={profileThing}>
        <SessionProvider>
          <PersonAvatar profileIri={profileIri} />
        </SessionProvider>
      </CombinedDataProvider>
    );
    const button = getByTestId(TESTCAFE_ID_UPLOAD_IMAGE);
    expect(button).toBeInTheDocument();
  });

  test("renders a confirmation dialog when delete pressed", async () => {
    const session = mockSession();
    const SessionProvider = mockSessionContextProvider(session);
    const { getByTestId } = renderWithTheme(
      <CombinedDataProvider solidDataset={profileDataset} thing={profileThing}>
        <SessionProvider>
          <PersonAvatar profileIri={profileIri} />
        </SessionProvider>
      </CombinedDataProvider>
    );

    const button = await screen.findByTestId(TESTCAFE_ID_REMOVE_IMAGE);
    await waitFor(() => {
      expect(button).toBeInTheDocument();
    });
    userEvent.click(button);

    const confirmationDialog = await screen.findByTestId(
      TESTCAFE_ID_CONFIRMATION_DIALOG
    );
    await waitFor(() => {
      expect(confirmationDialog).toBeInTheDocument();
    });

    expect(
      getByTestId(TESTCAFE_ID_CONFIRMATION_DIALOG_TITLE)
    ).toHaveTextContent(confirmationDialogTitle);
  });

  // change this to be more about user interaction
  test("confirmation dialog changes state as expected", async () => {
    const button = await screen.findByTestId(TESTCAFE_ID_REMOVE_IMAGE);
    await waitFor(() => {
      expect(button).toBeInTheDocument();
    });
    userEvent.click(button);
    // act(() => {
    //   openDeleteConfirmationDialogWrapper(mockDeleteFunc);
    // });
    // expect(PersonAvatar.title).toBe(confirmationDialogTitle);
    // expect(mockConfirmationDialog.open).toBeTruthy();
    // expect(deletePhotoFunction).toBe(mockDeleteFunc);
  });

  test("closes confirmation dialog when user clicks button", async () => {
    const mockDeleteFunc = jest.fn();
    const mockCloseDialog = jest.fn();
    const mockConfirmationDialog = mockConfirmationDialogContextProvider({
      closeDialog: mockCloseDialog,
    });
    act(() => {
      openDeleteConfirmationDialogWrapper(mockDeleteFunc);
    });
    expect(mockConfirmationDialog.closeDialog).toHaveBeenCalled();
  });
});

describe("setupErrorComponent", () => {
  it("renders", () => {
    const bem = (value) => value;
    const { asFragment } = render(setupErrorComponent(bem)());
    expect(asFragment()).toMatchSnapshot();
  });
});
