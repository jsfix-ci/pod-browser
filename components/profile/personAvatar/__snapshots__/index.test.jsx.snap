// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Person Avatar renders a person avatar 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <div
        class="PodBrowser-root PodBrowser-circle PodBrowser-avatar PodBrowser-colorDefault"
      >
        <svg
          aria-hidden="true"
          class="PodBrowser-root PodBrowser-fallback"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <input
        accept="image/*"
        class="PodBrowser-avatarInput"
        id="picture-upload-input-label"
        type="file"
      />
      <div
        class="PodBrowser-labelContainer"
        data-testid="profile-remove-image"
      >
        <svg
          aria-hidden="true"
          class="PodBrowser-root PodBrowser-removeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
        <label
          class="PodBrowser-inputLabelRemove"
        >
          Remove Photo
        </label>
      </div>
      <div
        class="PodBrowser-labelContainer"
      >
        <svg
          aria-hidden="true"
          class="PodBrowser-root PodBrowser-uploadIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM14 13v4h-4v-4H7l5-5 5 5h-3z"
          />
        </svg>
        <label
          class="PodBrowser-inputLabelUpload"
          data-testid="profile-upload-image"
          for="picture-upload-input-label"
        >
          Change Photo
        </label>
      </div>
    </div>
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <h3
        data-testid="profile-name-title"
      >
        <span
          class="PodBrowser-avatarText"
        >
          Alice
        </span>
        <a
          class="PodBrowser-headerLink"
          href="https://example.com/profile/card#me"
          rel="noreferrer"
          target="_blank"
        >
          http://example.com/webId#me
        </a>
      </h3>
    </div>
  </div>
</DocumentFragment>
`;

exports[`setupErrorComponent renders 1`] = `
<DocumentFragment>
  <div
    class="MuiAvatar-root MuiAvatar-circle avatar MuiAvatar-colorDefault"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiAvatar-fallback"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
